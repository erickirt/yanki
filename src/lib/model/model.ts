/* eslint-disable @typescript-eslint/naming-convention */

import { type YankiParamsForAction } from 'yanki-connect'

export type YankiModel = YankiParamsForAction<'createModel'>

export const yankiModels = [
	{
		cardTemplates: [
			{
				Back: '{{FrontSide}}\n\n<hr id=answer>\n\n{{Back}}',
				Front: '{{Front}}',
				YankiNamespace: '{{YankiNamespace}}',
			},
		],
		inOrderFields: ['Front', 'Back', 'YankiNamespace'],
		modelName: 'Yanki - Basic',
	},
	{
		// Changing the template structure slightly from the Anki defaults for
		// simplicity (instead of Text and Back Extra, we just have Back and Front)
		cardTemplates: [
			{
				Back: '{{cloze:Front}}<br>\n{{Back}}',
				Front: '{{cloze:Front}}',
				YankiNamespace: '{{YankiNamespace}}',
			},
		],
		inOrderFields: ['Front', 'Back', 'YankiNamespace'],
		isCloze: true,
		modelName: 'Yanki - Cloze',
	},
	{
		cardTemplates: [
			{
				Back: '{{Front}}\n\n<hr id=answer>\n\n{{type:Back}}',
				Front: '{{Front}}\n\n{{type:Back}}',
				YankiNamespace: '{{YankiNamespace}}',
			},
		],
		inOrderFields: ['Front', 'Back', 'YankiNamespace'],
		modelName: 'Yanki - Basic (type in the answer)',
	},
	{
		cardTemplates: [
			{
				Back: '{{FrontSide}}\n\n<hr id=answer>\n\n{{Back}}{{#Extra}}\n\n<hr>\n\n{{Extra}}{{/Extra}}',
				Front: '{{Front}}',
				YankiNamespace: '{{YankiNamespace}}',
			},
			{
				Back: '{{FrontSide}}\n\n<hr id=answer>\n\n{{Front}}{{#Extra}}\n\n<hr>\n\n{{Extra}}{{/Extra}}',
				Front: '{{Back}}',
				YankiNamespace: '{{YankiNamespace}}',
			},
		],
		inOrderFields: ['Front', 'Back', 'Extra', 'YankiNamespace'],
		modelName: 'Yanki - Basic (and reversed card with extra)',
	},
] as const satisfies YankiModel[]

export type YankiModelName = (typeof yankiModels)[number]['modelName']

export const yankiModelNames: YankiModelName[] = yankiModels.map((model) => model.modelName)
