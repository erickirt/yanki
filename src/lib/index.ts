export { type YankiNote } from './model/yanki-note'
export { getNoteFromMarkdown } from './parse/parse'
export { type CleanOptions, cleanNotes, defaultCleanOptions } from './sync/clean'
export { type ListOptions, defaultListOptions, listNotes } from './sync/list'
export { type SyncOptions, defaultSyncOptions, syncFiles, syncNotes } from './sync/sync'
export { hostAndPortToUrl, urlToHostAndPort } from './utilities/string'
