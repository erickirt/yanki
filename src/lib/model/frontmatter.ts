import { parse as yamlParse, stringify as yamlStringify } from 'yaml'

export type Frontmatter = {
	deckName?: string
	noteId?: number
	tags?: string[]
}

/**
 * Update the noteId in the frontmatter of a markdown string.
 *
 * Used when a noteId is received from Anki after creating a note.
 *
 *
 * String manipulation is ugly, but it ensures that the markdown format is
 * preserved verbatim. Running it through the unified AST and then
 * remarkStringify would possibly change the format.
 *
 * @param markdown Raw markdown string with frontmatter.
 * @param noteId  The value to set the noteId to. If undefined, the noteId will
 * be removed from the frontmatter. (Useful for testing.) @returns Raw markdown
 * string with updated frontmatter.
 */
export async function setNoteIdInFrontmatter(
	markdown: string,
	noteId: number | undefined,
): Promise<string> {
	const lines = markdown.split('\n')
	const frontmatterStart = lines.findIndex((line) => line.startsWith('---'))
	const frontmatterEnd = lines.findIndex(
		(line, index) => index > frontmatterStart && line.startsWith('---'),
	)

	if (frontmatterStart === -1 || frontmatterEnd === -1) {
		// No nothing if no noteId is provided
		if (noteId === undefined) {
			return markdown
		}

		// Add frontmatter with noteID
		const newFrontmatter = yamlStringify({ noteId }).trim()
		return ['---', newFrontmatter, '---\n', ...lines].join('\n')
	}

	const frontmatter = lines.slice(frontmatterStart + 1, frontmatterEnd).join('\n')
	const parsedFrontmatter = ((await yamlParse(frontmatter)) ?? {}) as Frontmatter

	if (noteId === undefined) {
		delete parsedFrontmatter.noteId
	} else {
		parsedFrontmatter.noteId = noteId
	}

	const newFrontmatter = yamlStringify(parsedFrontmatter).trim()

	return [
		...lines.slice(0, frontmatterStart + 1),
		newFrontmatter,
		...lines.slice(frontmatterEnd),
	].join('\n')
}
