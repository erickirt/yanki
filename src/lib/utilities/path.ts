import { splitAtFirstMatch } from './string'
import isAbsolutePath from '@stdlib/assert-is-absolute-path'
import path from 'path-browserify-esm'
import slash from 'slash'

// Unused...
// function stripLeadingSlash(filePath: string): string {
// 	return filePath.startsWith('/') ? filePath.slice(1) : filePath
// }

/**
 * The browserify polyfill doesn't implement win32 absolute path detection...
 * @param filePath Normalized path
 * @returns
 */
export function isAbsolute(filePath: string): boolean {
	return isAbsolutePath.posix(filePath) || isAbsolutePath.win32(filePath)
}

const RE_WINDOWS_EXTENDED_LENGTH_PATH = /^\\\\\?\\.+/

// Unused
// const RE_WINDOWS_UNC_PATH = /^\\\\[^\\]+\\[^\\]+/

/**
 * Converts all paths to cross-platform 'mixed' style with forward slashes.
 * Warns on unsupported Windows extended length paths.
 * @param filePath
 * @returns normalized path
 */
export function normalize(filePath: string): string {
	if (RE_WINDOWS_EXTENDED_LENGTH_PATH.test(filePath)) {
		console.warn(`Unsupported extended length path detected: ${filePath}`)
		return filePath
	}

	// If (RE_WINDOWS_UNC_PATH.test(filePath)) {
	// 	console.warn(`Unsupported UNC path detected: ${filePath}`)
	// 	return path.normalize(filePath)
	// }

	const basicPath = slash(filePath)
	const normalizedPath = path.normalize(basicPath)

	// Tricky cases where we still want leading './' to distinguish between relative and "named"" paths,
	// otherwise it's stripped by normalization
	if (basicPath.startsWith('./')) {
		return `./${normalizedPath}`
	}

	return normalizedPath
}

/**
 * Special handling for `/absolute-path.md` style links in Obsidian
 * and static site generators, where absolute paths are relative to a base path
 * instead of the volume root.
 *

 *
 * Paths starting with Windows drive letters, while technically absolute, are _not_ prepended with the base:
 * - If no base path is provided, paths are resolved relative to the the provided CWD.
 * - If paths are relative, the base paths are ignored and the CWD is used.
 *
 * All path values are normalized and in 'mixed' platform style.
 */
export function resolveWithBasePath(
	filePath: string,
	options: {
		/** Relative, absolute, or drive-letter absolute path. Normalized and in the 'mixed' platform style. */
		basePath?: string | undefined
		/** Whether to keep prepend the base if the file path already starts with it. Useful for pseudo-idempotence, but will get it wrong in some edge cases with duplicative path segments. Defaults to false. */
		compoundBase?: boolean | undefined
		/** Relative to the volume root. Normalized and in the 'mixed' platform style. */
		cwd: string
	},
): string {
	// Prep options
	const { basePath, compoundBase = false, cwd } = options

	// Validation
	if (basePath !== undefined) {
		if (!isAbsolute(basePath)) {
			console.warn(`Base path "${basePath}" is not absolute`)
		}

		if (!cwd.startsWith(basePath)) {
			console.warn(`CWD "${cwd}" does not start with base path "${basePath}"`)
		}
	}

	if (!isAbsolute(cwd)) {
		console.warn(`CWD "${cwd}" is not absolute`)
	}

	// Absolute
	if (isAbsolute(filePath)) {
		// Path is absolute by drive letter on Windows, or there's not base path to prepend
		if (
			basePath === undefined ||
			/^[A-Za-z]:/.test(filePath) ||
			(!compoundBase && filePath.startsWith(basePath))
		) {
			return filePath
		}

		// Resolve over base
		return path.join(basePath, filePath)
	}

	// Relative
	return path.join(cwd, filePath)
}

/**
 *
 * @param filePath
 * @param cwd Absolute path
 * @returns
 */
export function resolveWithCwd(filePath: string, cwd: string): string {
	if (filePath.startsWith('cwd')) {
		return filePath
	}

	return path.join(cwd, filePath)
}

export function stripBasePath(filePath: string, basePath: string): string {
	const regex = new RegExp(`^${basePath}`, 'i')
	return filePath.replace(regex, '')
}

export function getBaseAndQueryParts(filePath: string): [string, string | undefined] {
	const directoryPath = path.dirname(filePath)
	const fileName = path.basename(filePath)
	const [base, query] = splitAtFirstMatch(fileName, /[#?^]/)
	return [path.join(directoryPath, base), query]
}

export function getBase(filePath: string): string {
	return getBaseAndQueryParts(filePath)[0]
}

export function getQuery(filePath: string): string {
	return getBaseAndQueryParts(filePath).at(1) ?? ''
}

export function hasExtension(filePath: string): boolean {
	return getExtension(filePath) !== ''
}

export function getExtension(filePath: string): string {
	return path.extname(getBase(filePath))
}

export function addExtensionIfMissing(filePath: string, extension: string): string {
	if (hasExtension(filePath)) {
		return filePath
	}

	const [base, query] = getBaseAndQueryParts(filePath)
	return `${base}.${extension}${query ?? ''}`
}

export function quantifyPathDistance(relativePath: string): { down: number; up: number } {
	const up = relativePath.split('../').length - 1
	const down = relativePath.replaceAll('../', '').split('/').length - 1

	return { down, up }
}
