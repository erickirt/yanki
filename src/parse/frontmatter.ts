import type { Root } from 'mdast'
import { EXIT, visit } from 'unist-util-visit'
import { parse as yamlParse } from 'yaml'

export type Frontmatter = {
	deck?: string
	id?: number
	tags?: string[]
}

export function getFrontmatterFromTree(ast: Root): Frontmatter {
	let rawYaml: string | undefined
	visit(ast, 'yaml', (node) => {
		rawYaml = node.value
		return EXIT
	})

	if (rawYaml === undefined) {
		console.warn('No frontmatter found')
		return {}
	}

	return yamlParse(rawYaml) as Frontmatter
}
