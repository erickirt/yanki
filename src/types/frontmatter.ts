/**
 * Extract frontmatter from a markdown tree
 */

export type Frontmatter = {
	deck?: string
	noteId?: number
	tags?: string[]
}
