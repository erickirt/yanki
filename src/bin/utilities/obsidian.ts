// Node-only helpers for CLI

// Obsidian global settings:
// https://help.obsidian.md/Files+and+folders/How+Obsidian+stores+data#Global+settings

import fs from 'node:fs/promises'
import path from 'node:path'
import untildify from 'untildify'
import { normalize } from '../../lib/utilities/path'
import { PLATFORM } from '../../lib/utilities/platform'

type ObsidianVault = {
	directory: string
	id: string
	name: string
}

function getObsidianGlobalSettingsDirectory(): string {
	switch (PLATFORM) {
		case 'linux': {
			return normalize(path.join(untildify(process.env.XDG_CONFIG_HOME ?? '~/.config'), 'Obsidian'))
		}

		case 'mac': {
			return normalize(untildify('~/Library/Application Support/obsidian'))
		}

		case 'other': {
			throw new Error('Unsupported platform')
		}

		case 'windows': {
			return normalize(`${process.env.APPDATA}\\Obsidian`)
		}
	}
}

async function getObsidianVaults(): Promise<ObsidianVault[]> {
	const obsidianConfigFilePath = path.join(getObsidianGlobalSettingsDirectory(), 'obsidian.json')

	// Check if path exists
	try {
		await fs.access(obsidianConfigFilePath)
	} catch {
		return []
	}

	const obsidianConfig = JSON.parse(await fs.readFile(obsidianConfigFilePath, 'utf8')) as {
		vaults?: Record<string, { path: string }>
	}

	if (obsidianConfig.vaults === undefined) {
		return []
	}

	return Object.entries(obsidianConfig.vaults).map(([id, { path: directory }]) => ({
		directory,
		id,
		name: path.basename(directory),
	}))
}

/**
 * Assumes the vault has been opened at least once.
 * Searches up.
 */
export async function detectVault(fileOrDirectoryPath: string): Promise<ObsidianVault | undefined> {
	const obsidianVaults = await getObsidianVaults()

	if (obsidianVaults.length === 0) {
		return undefined
	}

	const normalizedPath = path.resolve(normalize(untildify(fileOrDirectoryPath)))

	// Check if directory
	const stats = await fs.stat(normalizedPath)

	const normalizedDirectoryPath = stats.isDirectory()
		? normalizedPath
		: path.dirname(normalizedPath)

	for (const vault of obsidianVaults) {
		const vaultDirectory = path.resolve(normalize(vault.directory))
		if (normalizedDirectoryPath.startsWith(vaultDirectory)) {
			return vault
		}
	}

	return undefined
}

// /**
//  * Primarily for testing.
//  * Node-only.
//  */
// export function openVault()
