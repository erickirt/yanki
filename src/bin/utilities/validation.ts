import { urlToHostAndPort } from '../../lib/utilities/url'

/**
 * Get host and port components of a URL
 *
 * @param url
 * @returns host and port
 * @throws if URL can't be parsed
 */
export function urlToHostAndPortValidated(url: string): { host: string; port: number } {
	const parsedUrl = urlToHostAndPort(url)

	if (parsedUrl === undefined) {
		throw new Error(`Invalid AnkiConnect URL: "${url}"`)
	}

	return parsedUrl
}
