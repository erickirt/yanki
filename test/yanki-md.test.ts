import { getNoteFromMarkdown, syncFiles } from '../src/lib'
import { getAllFrontmatter } from '../src/lib/model/frontmatter'
import { describeWithFileFixture } from './fixtures/file-fixture'
import { stableResults } from './utilities/stable-sync-results'
import fs from 'node:fs/promises'
import path from 'node:path'
import sortKeys from 'sort-keys'
import { expect, it } from 'vitest'

describeWithFileFixture(
	'model types',
	{
		assetPath: './test/assets/test-minimal-notes/',
		cleanUpAnki: true,
	},
	(context) => {
		it('correctly infers Anki model types from markdown', async () => {
			const results: Record<string, string> = {}
			for (const filePath of context.files) {
				const markdown = await fs.readFile(filePath, 'utf8')
				const { modelName } = await getNoteFromMarkdown(markdown, {
					namespace: context.namespace,
				})
				results[path.basename(filePath)] = modelName
			}

			expect(sortKeys(results, { deep: true })).toMatchInlineSnapshot(`
				{
				  "basic-and-reversed-card-with-no-back.md": "Yanki - Basic (and reversed card)",
				  "basic-and-reversed-card-with-no-front.md": "Yanki - Basic (and reversed card)",
				  "basic-and-reversed-card.md": "Yanki - Basic (and reversed card)",
				  "basic-type-in-the-answer-with-empty-frontmatter.md": "Yanki - Basic (type in the answer)",
				  "basic-type-in-the-answer-with-frontmatter.md": "Yanki - Basic (type in the answer)",
				  "basic-type-in-the-answer-with-multiple-emphasis-and-ignored-answer-style.md": "Yanki - Basic (type in the answer)",
				  "basic-type-in-the-answer-with-multiple-emphasis.md": "Yanki - Basic (type in the answer)",
				  "basic-type-in-the-answer.md": "Yanki - Basic (type in the answer)",
				  "basic-with-back-and-no-front-with-empty-frontmatter.md": "Yanki - Basic",
				  "basic-with-back-and-no-front.md": "Yanki - Basic",
				  "basic-with-cloze-like-back-and-no-front.md": "Yanki - Basic",
				  "basic-with-empty-everything.md": "Yanki - Basic",
				  "basic-with-empty-frontmatter.md": "Yanki - Basic",
				  "basic-with-front-and-cloze-like-back.md": "Yanki - Basic",
				  "basic-with-front-and-no-back.md": "Yanki - Basic",
				  "basic-with-type-in-the-answer-like-back-and-no-front.md": "Yanki - Basic",
				  "basic-with-type-in-the-answer-like-front-and-no-back.md": "Yanki - Basic",
				  "basic-with-type-in-the-answer-like-single-line-with-empty-frontmatter.md": "Yanki - Basic",
				  "basic-with-type-in-the-answer-like-single-line-with-frontmatter.md": "Yanki - Basic",
				  "basic-with-type-in-the-answer-like-single-line.md": "Yanki - Basic",
				  "basic.md": "Yanki - Basic",
				  "cloze-with-extra-empty.md": "Yanki - Cloze",
				  "cloze-with-extra.md": "Yanki - Cloze",
				  "cloze-with-style.md": "Yanki - Cloze",
				  "cloze.md": "Yanki - Cloze",
				}
			`)
		})
	},
)

describeWithFileFixture(
	'basic synchronization',
	{
		assetPath: './test/assets/test-minimal-notes/',
		cleanUpAnki: true,
	},
	(context) => {
		it('synchronizes notes to anki and has he correct deck name', async () => {
			const results = await syncFiles(context.files, { namespace: context.namespace })

			// Check the stuff that's elided from the stable results snapshot
			expect(results.duration).toBeDefined()
			expect(results.duration).toBeGreaterThan(0)

			for (const synced of results.synced) {
				expect(synced.note.noteId).toBeDefined()
				expect(synced.note.noteId).toBeGreaterThan(0)

				expect(context.files).toContain(synced.filePath)
			}

			expect(stableResults(results)).toMatchSnapshot()

			const deckNames = results.synced.map((syncInfo) => syncInfo.note.deckName)
			for (const deckName of deckNames) {
				expect(deckName).toBe(path.basename(context.assetPath))
			}
		})

		it('writes anki note IDs to the markdown files frontmatter', async () => {
			for (const filePath of context.files) {
				const markdown = await fs.readFile(filePath, 'utf8')

				const note = await getNoteFromMarkdown(markdown, { namespace: context.namespace })

				expect(note.noteId).toBeDefined()
				expect(note.noteId).toBeGreaterThan(0)
			}
		})
	},
)

describeWithFileFixture(
	'surplus frontmatter',
	{
		assetPath: './test/assets/test-surplus-frontmatter/',
		cleanUpAnki: true,
	},
	(context) => {
		it('preserves and merges unrelated surplus frontmatter', async () => {
			const results = await syncFiles(context.files, { namespace: context.namespace })
			expect(stableResults(results)).toMatchSnapshot()

			for (const filePath of context.files) {
				const markdown = await fs.readFile(filePath, 'utf8')
				const note = await getNoteFromMarkdown(markdown, { namespace: context.namespace })
				expect(note.noteId).toBeDefined()
				expect(note.noteId).toBeGreaterThan(0)

				const allFrontmatter = await getAllFrontmatter(markdown)

				expect(allFrontmatter['more stuff']).toBeDefined()
				expect(allFrontmatter['more stuff']).toBeTypeOf('object')
				expect(allFrontmatter['more stuff']).toMatchSnapshot()
				expect(allFrontmatter.noteId).toBeDefined()
				expect(allFrontmatter.noteId).toBeGreaterThan(0)
				expect(allFrontmatter.something).toBeDefined()
				expect(allFrontmatter.something).toMatchSnapshot()
			}
		})
	},
)

describeWithFileFixture(
	'complex trees',
	{
		assetPath: './test/assets/test-complex-tree/',
		cleanUpAnki: true,
	},
	(context) => {
		it('makes the right decisions about deck naming', async () => {
			const results = await syncFiles(context.files, { namespace: context.namespace })

			// Log inline for legibility
			const pathToDeckMap: Record<string, string | undefined> = {}
			for (const synced of results.synced) {
				const cleanPath =
					synced.filePath === undefined
						? `(Note is in Anki, no file path.)`
						: `/${path.basename(context.assetPath)}${synced.filePath.split(path.basename(context.assetPath), 2).pop() ?? ''}`
				pathToDeckMap[cleanPath] = synced.note.deckName
			}

			expect(pathToDeckMap).toMatchInlineSnapshot(`
				{
				  "/test-complex-tree/deep-contiguous/basic.md": "test-complex-tree::deep-contiguous",
				  "/test-complex-tree/deep-contiguous/within/basic.md": "test-complex-tree::deep-contiguous::within",
				  "/test-complex-tree/deep-contiguous/within/within/basic.md": "test-complex-tree::deep-contiguous::within::within",
				  "/test-complex-tree/deep-contiguous/within/within/within/basic.md": "test-complex-tree::deep-contiguous::within::within::within",
				  "/test-complex-tree/deep-island/within/within/within/basic.md": "test-complex-tree::deep-island::within::within::within",
				  "/test-complex-tree/deep-non-contiguous/within/within/basic.md": "test-complex-tree::deep-non-contiguous::within::within",
				  "/test-complex-tree/sibling-folders/brother/basic.md": "test-complex-tree::sibling-folders::brother",
				  "/test-complex-tree/sibling-folders/sister/basic.md": "test-complex-tree::sibling-folders::sister",
				  "/test-complex-tree/solo-note/basic.md": "test-complex-tree::solo-note",
				}
			`)

			expect(stableResults(results)).toMatchSnapshot()
		})
	},
)

describeWithFileFixture(
	'fancy markdown',
	{
		assetPath: './test/assets/test-fancy-markdown/',
		cleanUpAnki: true,
	},
	(context) => {
		it('handles fancy markdown', async () => {
			const results = await syncFiles(context.files, {
				namespace: context.namespace,
				obsidianVault: 'Vault',
			})

			expect(stableResults(results)).toMatchSnapshot()
		})
	},
)

describeWithFileFixture(
	'idempotent syncing',
	{
		assetPath: './test/assets/test-minimal-notes/',
		cleanUpAnki: true,
	},
	(context) => {
		it('idempotent syncing', async () => {
			// First sync should be all "created"
			const results = await syncFiles(context.files, {
				namespace: context.namespace,
				obsidianVault: 'Vault',
			})

			const syncActions = [...new Set(results.synced.map((syncInfo) => syncInfo.action))]

			expect(syncActions).toMatchInlineSnapshot(`
				[
				  "created",
				]
			`)

			const secondSyncResults = await syncFiles(context.files, {
				namespace: context.namespace,
				obsidianVault: 'Vault',
			})

			const secondSyncActions = [
				...new Set(secondSyncResults.synced.map((syncInfo) => syncInfo.action)),
			]

			expect(secondSyncActions).toMatchInlineSnapshot(`
				[
				  "unchanged",
				]
			`)
		})
	},
)
