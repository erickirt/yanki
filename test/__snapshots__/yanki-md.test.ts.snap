// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic synchronization > synchronizes notes to anki and has he correct deck name 1`] = `
{
  "deletedDecks": [],
  "deletedNotes": [],
  "duration": 0,
  "synced": [
    {
      "action": "created",
      "filePath": "basic-and-reversed-card.md",
      "note": {
        "deckName": "minimal-notes",
        "fields": {
          "Back": "<p>I'm an <strong>answer</strong> which is sometimes the question</p>
",
          "Front": "<p>I'm <em>question</em> which is sometimes the answer</p>
",
          "YankiNamespace": "Yanki Basic Sync Test",
        },
        "modelName": "Yanki - Basic (and reversed card)",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic-no-back.md",
      "note": {
        "deckName": "minimal-notes",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm the front of the card</p>
<p>I have no back.</p>
",
          "YankiNamespace": "Yanki Basic Sync Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic-type-in-the-answer.md",
      "note": {
        "deckName": "minimal-notes",
        "fields": {
          "Back": "I'm the thing you need to type on the card",
          "Front": "<p>I'm the prompt</p>
<h1>What an amazing</h1>
<ul>
<li><em>thing</em></li>
<li>that</li>
<li>is</li>
</ul>
",
          "YankiNamespace": "Yanki Basic Sync Test",
        },
        "modelName": "Yanki - Basic (type in the answer)",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "minimal-notes",
        "fields": {
          "Back": "<p>I'm the back of the card</p>
",
          "Front": "<p>I'm the front of the card</p>
",
          "YankiNamespace": "Yanki Basic Sync Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "cloze-extra.md",
      "note": {
        "deckName": "minimal-notes",
        "fields": {
          "Back": "<p>And some extra stuff.</p>
",
          "Front": "<p>This card has a {{c1::cloze}} or {{c1::two::(here's a hint)}}.</p>
",
          "YankiNamespace": "Yanki Basic Sync Test",
        },
        "modelName": "Yanki - Cloze",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "cloze.md",
      "note": {
        "deckName": "minimal-notes",
        "fields": {
          "Back": "",
          "Front": "<p>This card has a {{c1::cloze}} or {{c1::two::(here's a hint)}}.</p>
",
          "YankiNamespace": "Yanki Basic Sync Test",
        },
        "modelName": "Yanki - Cloze",
        "noteId": 0,
        "tags": [],
      },
    },
  ],
}
`;

exports[`complex trees > makes the right decisions about deck naming 2`] = `
{
  "deletedDecks": [],
  "deletedNotes": [],
  "duration": 0,
  "synced": [
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::deep-contiguous",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm just a note, you know?</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::solo-note",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm just a note, you know?</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::deep-contiguous::within",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I should be in the deck 'another'</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [
          "mdank",
          "yes",
        ],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::sibling-folders::brother",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I should be in the deck 'another'</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [
          "mdank",
          "yes",
        ],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::sibling-folders::sister",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I should be in the deck 'another'</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [
          "mdank",
          "yes",
        ],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::deep-contiguous::within::within",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I should be in the deck 'another'</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [
          "mdank",
          "yes",
        ],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::deep-non-contiguous::within::within",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm just a note, you know?</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::deep-contiguous::within::within::within",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I should be in the deck 'another'</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [
          "mdank",
          "yes",
        ],
      },
    },
    {
      "action": "created",
      "filePath": "basic.md",
      "note": {
        "deckName": "complex-tree::deep-island::within::within::within",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm just a note, you know?</p>
",
          "YankiNamespace": "Yanki Complex Tree Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
  ],
}
`;

exports[`surplus frontmatter > preserves and merges unrelated surplus frontmatter 1`] = `
{
  "deletedDecks": [],
  "deletedNotes": [],
  "duration": 0,
  "synced": [
    {
      "action": "created",
      "filePath": "basic-with-irrelevent-frontmatter.md",
      "note": {
        "deckName": "surplus-frontmatter",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm a note</p>
",
          "YankiNamespace": "Yanki Surplus Frontmatter Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
    {
      "action": "created",
      "filePath": "basic-with-surplus-frontmatter.md",
      "note": {
        "deckName": "surplus-frontmatter",
        "fields": {
          "Back": "<p><em>Intentionally blank.</em></p>
",
          "Front": "<p>I'm a note</p>
",
          "YankiNamespace": "Yanki Surplus Frontmatter Test",
        },
        "modelName": "Yanki - Basic",
        "noteId": 0,
        "tags": [],
      },
    },
  ],
}
`;

exports[`surplus frontmatter > preserves and merges unrelated surplus frontmatter 2`] = `
[
  "yes",
  "no",
  "maybe",
]
`;

exports[`surplus frontmatter > preserves and merges unrelated surplus frontmatter 3`] = `"unrelated"`;

exports[`surplus frontmatter > preserves and merges unrelated surplus frontmatter 4`] = `
[
  "yes",
  "no",
  "maybe",
]
`;

exports[`surplus frontmatter > preserves and merges unrelated surplus frontmatter 5`] = `"unrelated"`;
